safe_control_gym/controllers/mpc/mpc.py

```diff
@@ -216,14 +216,17 @@ def compute_initial_guess(self, init_state, goal_states=None):
                 u = self.lqr_gain @ (x_guess[:, i] - goal_states[:, i]) + np.atleast_2d(self.model.U_EQ)[0, :].T
                 u_guess[:, i] = u
                 x_guess[:, i + 1, None] = self.dynamics_func(x0=x_guess[:, i], p=u)['xf'].toarray()
+        elif self.compute_initial_guess_method == 'heuristic':
+            x_guess = goal_states
+            u_guess = np.repeat(np.atleast_2d(self.U_EQ), self.T, axis=0).T
         else:
             raise Exception('Initial guess method not implemented.')
 
         self.x_prev = x_guess
         self.u_prev = u_guess
 
         # set the solver back
-        self.setup_optimizer(solver=self.solver)
+        self.setup_optimizer(solver=self.solver)  # TODO why? this method is also called from mpc_acados! at least move to line 209!
```

Comment:
With my comment I wanted to highlight that when you call `compute_initial_guess()` from mpc_acados.py you end up calling  `compute_initial_guess()`from mpc.py and also you run line 229 which creates an instance of a casadi nlpsolver that is not needed in mpcacados.py.
Line 229 is important only in case you are using mpc.py, in order to overwrite the solver at line 188.

Reply:
Hmmm I see your point, we're setting up casadi for no reason. We should fix this, will talk to the group

Reply:
This was added by me. This idea of this is allowing solver swiching for mpc. For example, do an initial guess (`self.init_solver`) with ipopt then switch to a faster sqp solver (`self.solver`). I have tested with cartpole previously and this can improve the overall runtime and accuracy. But indeed acados will not use this. I think we can keep this if there is no clear interference with other code. 

One problem I notice is that this will add several milliseconds to the runtime. So the init step runtime needs a special treatment (maybe simply remove the first step).